### PR TITLE
PHOENIX-6632 Migrate/Update connectors to spark-3

### PR DESCRIPTION
**Pull request by ashwinb1998**
With Spark-3, the DatasourceV2 API has had major changes, where a new TableProvider Interface has been introduced. These new changes bring in more control to the data source developer and better integration with spark-optimizer.

For building: 
```
mvn clean package -Dphoenix-version=5.1.2 -Dhadoop-three.version=3.2.1 -Dhbase-two.version=2.4.4 -Dspark.version=3.1.2 -Dhbase-thirdparty-version=3.4.1 -Dscala.version=2.12.10 -Dscala.binary.version=2.12 -DskipTests -Dskip-phoenix4

```

### CODE REVIEW

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   5m 30s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 18 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |  11m 50s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  16m 37s |  master passed  |
| +1 :green_heart: |  compile  |   2m 35s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 43s |  master passed  |
| +1 :green_heart: |  javadoc  |   1m 13s |  master passed  |
| -1 :x: |  scaladoc  |   0m 10s |  root in master failed.  |
| +0 :ok: |  spotbugs  |   2m 12s |  phoenix-spark-base in master has 40 extant spotbugs warnings.  |
| +0 :ok: |  spotbugs  |   3m 44s |  root in master has 116 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 28s |  Maven dependency ordering for patch  |
| -1 :x: |  mvninstall  |   9m 55s |  root in the patch failed.  |
| -1 :x: |  compile  |   1m 37s |  root in the patch failed.  |
| -1 :x: |  javac  |   1m 37s |  root in the patch failed.  |
| -1 :x: |  scalac  |   1m 37s |  root in the patch failed.  |
| -1 :x: |  checkstyle  |   0m 50s |  root: The patch generated 262 new + 743 unchanged - 169 fixed = 1005 total (was 912)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  xml  |   0m  3s |  The patch has no ill-formed XML file.  |
| -1 :x: |  javadoc  |   0m 20s |  phoenix-spark-base generated 180 new + 4 unchanged - 6 fixed = 184 total (was 10)  |
| -1 :x: |  javadoc  |   1m  1s |  root generated 180 new + 14 unchanged - 6 fixed = 194 total (was 20)  |
| -1 :x: |  scaladoc  |   0m 39s |  phoenix-spark-base in the patch failed.  |
| -1 :x: |  scaladoc  |   0m  9s |  root in the patch failed.  |
| -1 :x: |  spotbugs  |   0m 40s |  phoenix-spark-base in the patch failed.  |
| -1 :x: |  spotbugs  |   2m 49s |  root in the patch failed.  |
||| _ Other Tests _ |
| -1 :x: |  unit  |  50m 44s |  root in the patch failed.  |
| -1 :x: |  asflicense  |   0m 18s |  The patch generated 5 ASF License warnings.  |
|  |   | 116m  7s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix-connectors/pull/69 |
| Optional Tests | dupname asflicense javac javadoc unit xml compile spotbugs hbaseanti checkstyle scalac scaladoc |
| uname | Linux b0afb2124731 4.15.0-156-generic #163-Ubuntu SMP Thu Aug 19 23:31:58 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-connectors-personality.sh |
| git revision | master / 3bf822f |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| scaladoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/branch-scaladoc-root.txt |
| mvninstall | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/patch-mvninstall-root.txt |
| compile | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/patch-compile-root.txt |
| javac | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/patch-compile-root.txt |
| scalac | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/patch-compile-root.txt |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/diff-checkstyle-root.txt |
| javadoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/diff-javadoc-javadoc-phoenix-spark-base.txt |
| javadoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/diff-javadoc-javadoc-root.txt |
| scaladoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/patch-scaladoc-phoenix-spark-base.txt |
| scaladoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/patch-scaladoc-root.txt |
| spotbugs | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/patch-spotbugs-phoenix-spark-base.txt |
| spotbugs | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/patch-spotbugs-root.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/patch-unit-root.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/testReport/ |
| asflicense | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/artifact/yetus-general-check/output/patch-asflicense-problems.txt |
| Max. process+thread count | 1909 (vs. ulimit of 30000) |
| modules | C: phoenix-spark-base . U: . |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/1/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Review by stoty — CHANGES_REQUESTED** [review]
Thanks a lot for the patch!

I have added comments for some issues, but overall it looks great.

The only major issue is with the packaging.
We support both Spark 1.6 for Phoenix 4.x, and we want to keep supporting Spark 2.4 for 5.0 in parallel with 3.0.

We have some quite ugly hacks that copy and overwrite the code between the 4.x and 5.x modules, and something similar should be done for new Spark3 connector.

However, first we should polish up the patch,  we can worry about the packaging issues later.

**Comment by ashwinb1998** [comment]
Thank you so much for you comments. I will work on these changes that you have mentioned.

Regarding the packaging, I was taking a look into the Apache Hudi package and they have separate [modules](https://github.com/apache/hudi/tree/master/hudi-spark-datasource/) for spark-2 and spark-3. The default spark-version is set to spark-2, but if someone wants to build with spark-3 they can do that as well. 

Is that something which can be done here as well?

File: phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/writer/PhoenixBatchWrite.java [hunk 1]
```diff
@@ -17,45 +17,35 @@
...
-    @Override
-    public DataWriterFactory<InternalRow> createWriterFactory() {
-        return new PhoenixDataWriterFactory(options);
+    PhoenixBatchWrite(LogicalWriteInfo writeInfo, Map<String,String> options) {
+        this.writeInfo = writeInfo;
+        this.options = createPhoenixDataSourceWriteOptions(options, writeInfo.schema());
     }
 
     @Override
-    public boolean useCommitCoordinator() {
```
[the remark this answers is not in the document]
Comment (ashwinb1998) [reply phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/writer/PhoenixBatchWrite.java:57]: It basically returns whether Spark should use the commit coordinator to ensure that at most one task for each partition commits. It has now been removed with DSV2 in spark3.

File: phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/writer/PhoenixBatchWrite.java [hunk 2]
```diff
@@ -17,45 +17,35 @@
...
 import static org.apache.phoenix.spark.datasource.v2.PhoenixDataSource.extractPhoenixHBaseConfFromOptions;
 
-public class PhoenixDataSourceWriter implements DataSourceWriter {
+public class PhoenixBatchWrite implements BatchWrite {
 
     private final PhoenixDataSourceWriteOptions options;
+    private final LogicalWriteInfo writeInfo;
 
-    public PhoenixDataSourceWriter(SaveMode mode, StructType schema, DataSourceOptions options) {
-        if (!mode.equals(SaveMode.Overwrite)) {
```
[the remark this answers is not in the document]
Comment (ashwinb1998) [reply phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/writer/PhoenixBatchWrite.java:39]: SaveMode has been removed from the DataSourceV2 API and hence the there is no parameter SaveMode that can be used here. Reference: https://issues.apache.org/jira/browse/SPARK-25531
It is a requirement though, that for the writes, SaveMode should be "Append".

File: phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/reader/PhoenixScan.java [hunk 3]
```diff
@@ -29,134 +29,97 @@
...
                 throw new NullPointerException();
             }
-
             final PhoenixStatement pstmt = statement.unwrap(PhoenixStatement.class);
             // Optimize the query plan so that we potentially use secondary indexes
             final QueryPlan queryPlan = pstmt.optimizeQuery(selectStatement);
-            final Scan scan = queryPlan.getContext().getScan();
-
-            // setting the snapshot configuration
-            Optional<String> snapshotName = options.get(PhoenixConfigurationUtil.SNAPSHOT_NAME_KEY);
```
[the remark this answers is not in the document]
Comment (ashwinb1998) [reply phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/reader/PhoenixScan.java:156]: Woops, somehow missed it, will add it now.

File: phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/reader/PhoenixDataSourceReadOptions.java [hunk 4]
```diff
@@ -57,6 +59,14 @@ String getTenantId() {
     }
 
     Properties getOverriddenProps() {
```
[the remark this answers is not in the document]
Comment (ashwinb1998) [reply phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/reader/PhoenixDataSourceReadOptions.java:61]: I have modified the way this was implemented, I went back to how it is currently.

File: phoenix-spark-base/pom.xml [hunk 5]
```diff
@@ -121,7 +121,7 @@
     <dependency>
       <groupId>org.scalatest</groupId>
       <artifactId>scalatest_${scala.binary.version}</artifactId>
-      <version>2.2.4</version>
+      <version>3.1.0</version>
```
[the remark this answers is not in the document]
Comment (ashwinb1998) [reply phoenix-spark-base/pom.xml:124]: No specific reason, I can change this to latest version if you'd think that would be better.

**Review by stoty — COMMENTED** [review]
This looks good, I've left comments for some minor issues.

On to the packaging issue:

Instead of changing the existing module, we should leave that alone, and add a new phoenix5-spark3 module.

The same goes for the spark version, add a new property for the spark3 version. - Now I see that it is specified directly, so it's not an issue.

I think that we should not use the org.apache.phoenix.spark.datasource.v2 package for the new module, but rather something like org.apache.phoenix.spark.sql.connector.

The biggest problem is currently the phoenix5 and phoenix4 spark connectors use the same version and the same code, while the spark3 connector changes that.

 This can be handled either by adding a completely new phoenix-spark3 top level module, or by having a base source tree for all versions, and merging it with the spark2 and spark3 specific files dring build.
 
Considering the amount of change reuqired for Spark3, and how ugly the merging solution is, I lean towards creating a completely new top level module.
 
 I don't know how experienced you are with maven, if you'd like, I can take over the module duplicating part once the code is done. (Which should be very soon)

File: phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/reader/PhoenixScan.java [hunk 6]
```diff
@@ -29,134 +29,104 @@
...
 
             // setting the snapshot configuration
-            Optional<String> snapshotName = options.get(PhoenixConfigurationUtil.SNAPSHOT_NAME_KEY);
-            if (snapshotName.isPresent())
-                PhoenixConfigurationUtil.setSnapshotNameKey(queryPlan.getContext().getConnection().
-                        getQueryServices().getConfiguration(), snapshotName.get());
+            String snapshotName = options.get(PhoenixConfigurationUtil.SNAPSHOT_NAME_KEY);
+
+            if(snapshotName != null){
+                PhoenixConfigurationUtil.setSnapshotNameKey(queryPlan.getContext().getConnection().getQueryServices().getConfiguration(), snapshotName);
```
[the remark this answers is not in the document]
Comment (stoty) [reply phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/reader/PhoenixScan.java:128]: It's not off, it's just the GitHUb formatting.

File: phoenix-spark-base/src/it/scala/org/apache/phoenix/spark/PhoenixSparkIT.scala [hunk 7]
```diff
@@ -48,7 +48,7 @@ class PhoenixSparkIT extends AbstractPhoenixSparkIT {
       .format("phoenix")
       .options(Map("table" -> "TABLE3",
         PhoenixDataSource.ZOOKEEPER_URL -> quorumAddress, PhoenixDataSource.SKIP_NORMALIZING_IDENTIFIER -> "true"))
-      .mode(SaveMode.Overwrite)
+      .mode(SaveMode.Append)
```
[the remark this answers is not in the document]
Comment (ashwinb1998) [reply phoenix-spark-base/src/it/scala/org/apache/phoenix/spark/PhoenixSparkIT.scala:51]: I believe there is an internal file source v2 API that uses Savemode for writes, which is why we need to specify `mode(SaveMode.Append)`. SaveMode is removed only from DataSourceV2 API, which means we won't be able to check from the parameter if the given condition is Append or Overwrite.

File: phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/PhoenixDataSource.java [hunk 8]
```diff
@@ -17,42 +17,76 @@
...
     public static final String ZOOKEEPER_URL = "zkUrl";
     public static final String PHOENIX_CONFIGS = "phoenixconfigs";
+    protected StructType schema;
+    private CaseInsensitiveStringMap options;
 
     @Override
-    public DataSourceReader createReader(DataSourceOptions options) {
-        return new PhoenixDataSourceReader(options);
+    public StructType inferSchema(CaseInsensitiveStringMap options){
+        if (options.get("table") == null) {
```
[the remark this answers is not in the document]
Comment (stoty) [reply phoenix-spark-base/src/main/java/org/apache/phoenix/spark/datasource/v2/PhoenixDataSource.java]: I can still find a lot of "table" strings in the code.
It should only have instance, where the TABLE_KEY constant is defined.

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   5m 33s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  1s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 18 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |  12m 31s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  16m 42s |  master passed  |
| +1 :green_heart: |  compile  |   2m 38s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 44s |  master passed  |
| +1 :green_heart: |  javadoc  |   1m 13s |  master passed  |
| -1 :x: |  scaladoc  |   0m 10s |  root in master failed.  |
| +0 :ok: |  spotbugs  |   2m 14s |  phoenix-spark-base in master has 40 extant spotbugs warnings.  |
| +0 :ok: |  spotbugs  |   3m 41s |  root in master has 116 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 38s |  Maven dependency ordering for patch  |
| -1 :x: |  mvninstall  |   9m 48s |  root in the patch failed.  |
| -1 :x: |  compile  |   1m 36s |  root in the patch failed.  |
| -1 :x: |  javac  |   1m 36s |  root in the patch failed.  |
| -1 :x: |  scalac  |   1m 36s |  root in the patch failed.  |
| -1 :x: |  checkstyle  |   0m 51s |  root: The patch generated 260 new + 746 unchanged - 166 fixed = 1006 total (was 912)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  xml  |   0m  2s |  The patch has no ill-formed XML file.  |
| -1 :x: |  javadoc  |   0m 22s |  phoenix-spark-base generated 180 new + 4 unchanged - 6 fixed = 184 total (was 10)  |
| -1 :x: |  javadoc  |   1m  0s |  root generated 180 new + 14 unchanged - 6 fixed = 194 total (was 20)  |
| -1 :x: |  scaladoc  |   0m 38s |  phoenix-spark-base in the patch failed.  |
| -1 :x: |  scaladoc  |   0m  8s |  root in the patch failed.  |
| -1 :x: |  spotbugs  |   0m 41s |  phoenix-spark-base in the patch failed.  |
| -1 :x: |  spotbugs  |   2m 45s |  root in the patch failed.  |
||| _ Other Tests _ |
| -1 :x: |  unit  |  65m 33s |  root in the patch failed.  |
| -1 :x: |  asflicense  |   0m 19s |  The patch generated 5 ASF License warnings.  |
|  |   | 131m 47s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix-connectors/pull/69 |
| Optional Tests | dupname asflicense javac javadoc unit xml compile spotbugs hbaseanti checkstyle scalac scaladoc |
| uname | Linux cb92bcfd7755 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-connectors-personality.sh |
| git revision | master / 72496e5 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| scaladoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/branch-scaladoc-root.txt |
| mvninstall | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/patch-mvninstall-root.txt |
| compile | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/patch-compile-root.txt |
| javac | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/patch-compile-root.txt |
| scalac | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/patch-compile-root.txt |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/diff-checkstyle-root.txt |
| javadoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/diff-javadoc-javadoc-phoenix-spark-base.txt |
| javadoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/diff-javadoc-javadoc-root.txt |
| scaladoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/patch-scaladoc-phoenix-spark-base.txt |
| scaladoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/patch-scaladoc-root.txt |
| spotbugs | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/patch-spotbugs-phoenix-spark-base.txt |
| spotbugs | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/patch-spotbugs-root.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/patch-unit-root.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/testReport/ |
| asflicense | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/artifact/yetus-general-check/output/patch-asflicense-problems.txt |
| Max. process+thread count | 1920 (vs. ulimit of 30000) |
| modules | C: phoenix-spark-base . U: . |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/3/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by ashwinb1998** [comment]
Please let me know if there are more changes, I will try to finish as quick as I can. 

Regarding the packaging, I believe I'm not quite experienced with maven, especially since this change would require a lot of modifications. It would be best if you could take over for the maven part. But definitely let me know if I can help in any way. 

Thanks

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 57s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 18 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |  11m  8s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  16m 52s |  master passed  |
| +1 :green_heart: |  compile  |   2m 34s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 44s |  master passed  |
| +1 :green_heart: |  javadoc  |   1m 15s |  master passed  |
| -1 :x: |  scaladoc  |   0m 10s |  root in master failed.  |
| +0 :ok: |  spotbugs  |   2m 10s |  phoenix-spark-base in master has 40 extant spotbugs warnings.  |
| +0 :ok: |  spotbugs  |   3m 43s |  root in master has 116 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 37s |  Maven dependency ordering for patch  |
| -1 :x: |  mvninstall  |  10m  5s |  root in the patch failed.  |
| -1 :x: |  compile  |   1m 33s |  root in the patch failed.  |
| -1 :x: |  javac  |   1m 33s |  root in the patch failed.  |
| -1 :x: |  scalac  |   1m 33s |  root in the patch failed.  |
| -1 :x: |  checkstyle  |   0m 49s |  root: The patch generated 263 new + 743 unchanged - 169 fixed = 1006 total (was 912)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  xml  |   0m  2s |  The patch has no ill-formed XML file.  |
| -1 :x: |  javadoc  |   0m 21s |  phoenix-spark-base generated 180 new + 4 unchanged - 6 fixed = 184 total (was 10)  |
| -1 :x: |  javadoc  |   1m  1s |  root generated 180 new + 14 unchanged - 6 fixed = 194 total (was 20)  |
| -1 :x: |  scaladoc  |   0m 39s |  phoenix-spark-base in the patch failed.  |
| -1 :x: |  scaladoc  |   0m  9s |  root in the patch failed.  |
| -1 :x: |  spotbugs  |   0m 40s |  phoenix-spark-base in the patch failed.  |
| -1 :x: |  spotbugs  |   2m 47s |  root in the patch failed.  |
||| _ Other Tests _ |
| -1 :x: |  unit  |  45m 27s |  root in the patch failed.  |
| -1 :x: |  asflicense  |   0m 18s |  The patch generated 5 ASF License warnings.  |
|  |   | 106m  1s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix-connectors/pull/69 |
| Optional Tests | dupname asflicense javac javadoc unit xml compile spotbugs hbaseanti checkstyle scalac scaladoc |
| uname | Linux ac994ead4a13 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-connectors-personality.sh |
| git revision | master / 72496e5 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| scaladoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/branch-scaladoc-root.txt |
| mvninstall | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/patch-mvninstall-root.txt |
| compile | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/patch-compile-root.txt |
| javac | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/patch-compile-root.txt |
| scalac | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/patch-compile-root.txt |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/diff-checkstyle-root.txt |
| javadoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/diff-javadoc-javadoc-phoenix-spark-base.txt |
| javadoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/diff-javadoc-javadoc-root.txt |
| scaladoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/patch-scaladoc-phoenix-spark-base.txt |
| scaladoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/patch-scaladoc-root.txt |
| spotbugs | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/patch-spotbugs-phoenix-spark-base.txt |
| spotbugs | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/patch-spotbugs-root.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/patch-unit-root.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/testReport/ |
| asflicense | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/artifact/yetus-general-check/output/patch-asflicense-problems.txt |
| Max. process+thread count | 1904 (vs. ulimit of 30000) |
| modules | C: phoenix-spark-base . U: . |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-Connectors-PreCommit-GitHub-PR/job/PR-69/4/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by stoty** [comment]
Thank you @ashwinb1998 , I'll finish the build system integration (and I'm probably gonna change the package as well.)

**Comment by ashwinb1998** [comment]
> 

Awesome sounds good. 
Just a small question, how would I push my changes once you finish the packaging part? 
since you were going to add it to a different package like org.apache.phoenix.spark.sql.connector

**Comment by stoty** [comment]
Changes need to be pushed by a committer.
I'm going to push the change when it's done.
I'm going to ask a review for it from you first.

I'll leave you as the author of the patch, and add myself as a co-author if that's all right with you.

**Comment by ashwinb1998** [comment]
> Changes need to be pushed by a committer. I'm going to push the change when it's done. I'm going to ask a review for it from you first.
> 
> I'll leave you as the author of the patch, and add myself as a co-author if that's all right with you.

Oh okay, sounds great. Its all good with me 🙂

**Comment by stoty** [comment]
Can you close this PR @ashwinb1998 , so that we don't have two open PRs for this ?

**Comment by ashwinb1998** [comment]
Sure @stoty, will do

**Comment by rejeb** [comment]
Hi,
I've noticed a breaking change related to this PR and I haven't found any mention of it on documentation.
For tables with non default column family spark DataFrameReader columnName for columns blonging to the column family are of form => "columnFamily.columnName".
This is caused by the change of the manner that the columName is retrieved:
- In spark2 connector code the columnName is retreived using the method ColumnInfo.getDisplayName which returns the columnName without columnFamily.
- In this PR, the columnName is retreived using ColumnInfo.getColumnName which returns the name as "columnFamily.columnName".
This causes two problems:
- We must rename columns to add backtik as `columnFamily.columnName` for spark to be able to resolve DF columnName
- When doing a select on only "columnName" spark can't resolve it since to DF name is "columnFamily.columnName"

For DataFrameWriter, everything work as in spark2. I haven't been able to find out how but it happens that the behavior is different between Reader and Writer.

I don't know if this is an issue or not. But it would be great to have a flag to manage this behavior and also fix schema creation to add backticks for columnFamily prefixed columns.

**Pull request closed without merging**